test/RelayHub.test.js

```diff
@@ -335,6 +338,294 @@ contract('RelayHub', function ([_, relayOwner, relay, sender, other]) {  // esli
     });
   });
 
+  describe('penalizations', function () {
+    describe('penalizable behaviors', function () {
+      beforeEach('staking for relay', async function () {
+        await relayHub.stake(relay, time.duration.days(1), { value: ether('1') });
+      });
+
+      const encodedCallArgs = {
+        sender,
+        recipient: '0x1820b744B33945482C17Dc37218C01D858EBc714',
+        data: '0x1234',
+        fee: 10,
+        gasPrice: 50,
+        gasLimit: 1000000,
+        nonce: 0,
+      };
+
+      const relayCallArgs = {
+        gasPrice: 50,
+        gasLimit: 1000000,
+        nonce: 0,
+        privateKey: '6370fd033278c143179d81c5526140625662b8daa446c22ee2d73db3707e620c',
+      };
+
+      describe('repeated relay nonce', async function () {
+        it('penalizes transactions with same nonce and different data', async function () {
+          const txDataSigA = getDataAndSignature(encodeRelayCall(encodedCallArgs, relayCallArgs));
+          const txDataSigB = getDataAndSignature(encodeRelayCall(Object.assign(encodedCallArgs, { data: "0xabcd" }), relayCallArgs));
+
+          await relayHub.penalizeRepeatedNonce(txDataSigA.data, txDataSigA.signature, txDataSigB.data, txDataSigB.signature);
```

Comment:
you only check if the method reverts (or not), but not if it actually remove the relay, and reward the snitcher..

Reply:
Correct, these tests (the whole 'penalizable behaviors' `describe` block) test the different ways in which a relay can be penalized. The following describe ('relay state and reward', not the best name) tests in which states a relay can be penalized, and what happens if it is.

Looking at Mocha's output may help understand how the tests are laid out:
```
Conctract: RelayHub  
  penalizations
    penalizable behaviors
      repeated relay nonce
        ✓ penalizes transactions with same nonce and different data
        ✓ does not penalize transactions with same nonce and same data
        ✓ does not penalize transactions with different nonces
        ✓ does not penalize transactions with same nonce from different relays
      illegal call
        with pre-EIP155 signatures
          ✓ penalizes relay transactions to addresses other than RelayHub
          ✓ penalizes relay transactions to illegal RelayHub functions (stake)
          ✓ penalizes relay transactions to illegal RelayHub functions (penalize)
          ✓ does not penalize legal relay transactions
    relay state and reward
      with penalizable transaction
        with unstaked relay
          ✓ account cannot be penalized
          with staked relay
            ✓ relay can be penalized
            once penalized
              ✓ relay cannot be penalized again
            with registered relay
              ✓ relay can be penalized
              ✓ RelayRemoved event is emitted
              once penalized
                ✓ relay cannot be penalized again
              with removed relay
                ✓ relay can be penalized
                once penalized
                  ✓ relay cannot be penalized again
                with unstaked relay
                  ✓ relay cannot be penalized
```

Reply:
I'm not sure I see an added value in separating the penalization tests to different `describe` blocks. What's the meaning of having `penalizes transactions with same nonce and different data` pass yet `relay state and reward` -> `with penalizable transaction` -> `with staked relay` -> `relay can be penalized` fail? or vice versa? why not just check the state after each penalizable flow that doesn't revert?

Reply:
>What's the meaning of having penalizes transactions with same nonce and different data pass yet relay state and reward -> with penalizable transaction -> with staked relay -> relay can be penalized fail? 

That would mean that:
a) two transactions with same nonce and different data can be penalized
b) a staked relay cannot be penalized

This could happen if e.g. the `penalize` function disallowed penalizing unregistered relays. Each `describe` block is testing different things: _the behaviors_ that can trigger a penalization, and _in which states_ penalizations can happen.

Reply:
Ok, let me correct myself then:
What's the meaning of having `penalizes transactions with same nonce and different data` pass yet `relay state and reward` -> `with penalizable transaction` -> `with **registered** relay` -> `relay can be penalized` fail? or vice versa? why not just check the state after each penalizable flow that doesn't revert?

The point I was making is that checking that the penalize didn't revert alone, and then in a different test checking the state changes is unnecessary. In the above case, if  `penalizes transactions with same nonce and different data` passes yet  `with **registered** relay` -> `relay can be penalized` fails, the first test is misleading and actually doesn't have any significant meaning, and on the other hand, if the state changed per the second test, then `penalizes transactions with same nonce and different data` will always pass with no exception but a bug in evm implementation that would revert yet allow state changes...

Reply:
> the first test is misleading and actually doesn't have any significant meaning

That would simply mean that a registered relay cannot be penalized (but perhaps an unregistered or removed one can).

>then penalizes transactions with same nonce and different data will always pass with no exception

Yes, but there are multiple reasons why a relay may be penalized, and same nonce different data is only one of them.

To get something closer to 'complete' coverage, you'd need to check all possible penalizable behaviors, on all possible states. I'm not going for a full black-box testing approach, and since I know penalizable behaviours are independent of the state the relay is in, I chose to also split those in the tests.

That said, I don't really see a reason to block this PR over such a thing, considering these new tests _do_ cover scenarios which weren't covered before, and don't really add a maintenance burden.

Reply:
the only thing we're saying is that instead of checking just for revert or not, to add a `expectPenalized()` method, which takes a promise and verify that afterward, the relay indeed lost its stake and the reported got its share.
that method should be called on each of those tests (except, of course, the test that expectRevert)

Reply:
and BTW: don't "resolve" issues that fast. It took me some time to find this thread, as it disappeared

Reply:
Ah, I see what you mean, that's a good idea. I've added that to the tests, and reworked them a bit since there was some unnecessary nesting, it should read a bit easier now.